### ISSUE TITLE
Performance issue with translations of `languages` and `countries`

### ISSUE THREAD

**Issue by fritzmg** [contao/contao#6454]
### Affected version(s)

4.13, 5.2

### Description

In Contao **4.12** we introduced the `@contao.intl.locales` and `@contao.intl.countires` services with which we can use translated locales and countries instead of `$GLOBALS['TL_LANG']['LNG']` and `$GLOBALS['TL_LANG']['CNT']` respectively. This allowed us to remove the `languages` and `countries` translation files. However, this introduced a performance issue.

Previously after `cache:warmup` there would have been a cache file for these translations within

```
%kernel.cache_dir%/contao/languages/<language>/languages.php
%kernel.cache_dir%/contao/languages/<language>/countries.php
```

But now, since the translations are loaded from `@contao.intl.locales` (i.e. from PHP intl) instead, this cache file is missing. This means that for the `contao_languages` and `contao_countries` translation domain, the following code is always executed every time

https://github.com/contao/contao/blob/c4e90db304d88d8414def7b74586dc24ed80edfd/core-bundle/src/Resources/contao/library/Contao/System.php#L545-L564

whenever

```php
System::loadLanguageFile('languages')
```

is executed - which will also be executed down the line here for example:

https://github.com/contao/contao/blob/c4e90db304d88d8414def7b74586dc24ed80edfd/core-bundle/src/Intl/Locales.php#L212

Meaning every time the file system is scanned in all `contao` resource folders. 

This can be especially noticeable when ChangeLanguage is in use for example (as reported by @Toflar), as it needs to access translations of locales multiple times for the change language menu for example.

**Comment by aschempp** [comment]
What's the reason for re-loading the same language file over and over again? Shouldn't we just cache that this file has been loaded already like we do for DCAs?

**Comment by fritzmg** [comment]
DCAs are only ever loaded once. Contao translations are however populated into `$GLOBALS['TL_LANG']` every time, depending on the requested locale and the last requested locale. So when you execute

```php
$this->translator->trans('foo', [], 'contao_default', 'de')
```

`$GLOBALS['TL_LANG']` will be populated with the `de` translations. If you execute

```php
$this->translator->trans('foo', [], 'contao_default', 'de')
```

again, nothing will happen within `System::loadLanguageFile`, as `$GLOBALS['TL_LANG']` is already populated with the `de` language. But if you then execute

```php
$this->translator->trans('foo', [], 'contao_default', 'en')
```

`$GLOBALS['TL_LANG']` will be populated with the `en` translations. This happens with the ChangeLanguage extension, as previously mentioned.

But that's not the issue here. Usually the Contao translations are loaded from a static file cache here:

https://github.com/contao/contao/blob/c4e90db304d88d8414def7b74586dc24ed80edfd/core-bundle/src/Resources/contao/library/Contao/System.php#L538-L542

But as already mentioned, this cannot happen for the `languages` and `countries` Contao translations anymore, as they do not exist (unless you or an extension provided some).

**Comment by Toflar** [comment]
What if we just do `file_put_contents($strCacheDir . '/contao/languages/' . $strCreateLang . '/' . $strName . '.php'...` in this case for future requests? Would taht be possible? I mean what would the content be?

**Comment by fritzmg** [comment]
> I mean what would the content be?

Just `<?php`. Though I don't know how we can implement this generally. Because there can be arbitrary translation files and the `ContaoCacheWarmer` cannot know about any missing ones - unless we hardcode some, which is not very elegant.

// edit: e.g.

https://github.com/contao/contao/blob/c4e90db304d88d8414def7b74586dc24ed80edfd/core-bundle/src/Cache/ContaoCacheWarmer.php#L112-L152

```diff
        foreach ($this->locales as $language) {
            $processed = [];
            $files = $this->findLanguageFiles($language);

            foreach ($files as $file) {
                $name = substr($file->getBasename(), 0, -4);

                if (\in_array($name, $processed, true)) {
                    continue;
                }

                $processed[] = $name;

                $subfiles = $this->finder
                    ->findIn(Path::join('languages', $language))
                    ->files()
                    ->name("/^$name\\.(php|xlf)$/")
                ;

                try {
                    $dumper->dump(
                        iterator_to_array($subfiles),
                        Path::join('languages', $language, "$name.php"),
                        ['type' => $language]
                    );
                } catch (\OutOfBoundsException $e) {
                    continue;
                }
            }

+           // Force empty "languages" and "countries" cache (#6454)
+           foreach (['languages', 'countries'] as $name) {
+               if (\in_array($name, $processed, true)) {
+                   continue;
+               }
+
+               $dumper->dump([],  Path::join('languages', $language, "$name.php"), ['type' >= $language]);
+           }
        }
```

**Comment by Toflar** [comment]
No I mean not during the cache build but actually in `System::loadLanguageFile()`. This means the first call still misses the cache but after that it would be ready?

**Comment by fritzmg** [comment]
> No I mean not during the cache build but actually in System::loadLanguageFile(). This means the first call still misses the cache but after that it would be ready?

Within `System::loadLanguageFile` we currently only check if the last language that was loaded is the same as the currently requsted language. If not, then the language files are loaded again (either from the `prod` cache or loaded via the resource finder).

**Comment by Toflar** [comment]
Ah, we don't know what was found so we cannot cache the contents. But is that supported? I mean, right now it works if you do not build an internal cache for your `my_custom_stuff` language domain. Because it's loaded on every request. Which is fine in `APP_ENV=dev` I guess but why would you expect this to work in `prod`?

**Comment by aschempp** [comment]
has this been fixed by #6524 ?

**Comment by Toflar** [comment]
Yes :)
